### PR TITLE
algorithm conversions

### PR DESCRIPTION
**Pull request by neheb**
## Description

Removes several for loops in favor of STL algorithms.

### CODE REVIEW

File: src/builtin_disown.cpp [hunk 1]
```diff
@@ -101,9 +101,9 @@ int builtin_disown(parser_t &parser, io_streams_t &streams, wchar_t **argv) {
         }
 
         // Disown all target jobs
-        for (const auto &j : jobs) {
-            retval |= disown_job(cmd, parser, streams, j);
-        }
+        return std::accumulate(jobs.begin(), jobs.end(), retval, [&](int r, job_t *j) {
+            return r || disown_job(cmd, parser, streams, j);
```
Comment (krobelus) [inline src/builtin_disown.cpp:105]: Looks like you have uncovered a bug in this linter - this should use bitwise "or" to always disown all jobs.

```cpp
#include <stdio.h>
#include <vector>
#include <numeric>

int main() {
	std::vector<int> vs = {0, 1, 2};

	int r1 = 0;
	for (const auto &v : vs)
		r1 |= v;

	int r2 = std::accumulate(vs.begin(), vs.end(), 0, [&](int r, int v) {
		return r || v; // should be r | v
	});

	printf("%d\n", r1); // prints 3
	printf("%d\n", r2); // prints 1
}
```

Comment (neheb) [reply src/builtin_disown.cpp]: Good catch.

All the changes were done manually. This was a mistake on my part.

**Comment by mqudsi** [comment]
I'm not sure if I'm the only one or not, but while some of these are nice the majority increase the verbosity while simultaneously making it less readable. But maybe it's just me?

**Comment by faho** [comment]
>while some of these are nice the majority increase the verbosity while simultaneously making it less readable. But maybe it's just me?

Not just you. But I think I might kinda hate modern C++ style.

Is there any appreciable difference in behavior here? Any performance benefits? Anything that's easier to extend or understand?

Because "avoid for-loops at all costs" is definitely not my style.

**Comment by neheb** [comment]
accumulate generates slightly more efficient code: https://godbolt.org/z/nefgs7 as an example.

No real idea about the other ones. Maybe count_if as well.

They also happen to be constexpr in C++20. Probably not relevant here.

**Comment by krobelus** [comment]
I like the `any_of` one, the others are a bit awkward.

**Comment by neheb** [comment]
Eh. I only agree about the find_if ones.

**Comment by neheb** [comment]
Rebased.

**Comment by ridiculousfish** [comment]
Thank you for the PR and your patience in review. However I am going to decline this. I agree with krobelus that we should prioritize readability. Beyond that for loops have other nice properties, like that you can `return` from inside them.

**Pull request closed without merging**